packages/agent-builder/src/workflows/template-builder.ts

```diff
@@ -0,0 +1,1682 @@
+      const packageJson = JSON.parse(packageJsonContent);
+
+      console.log('Template package.json:', JSON.stringify(packageJson, null, 2));
+
+      return {
+        dependencies: packageJson.dependencies || {},
+        devDependencies: packageJson.devDependencies || {},
+        peerDependencies: packageJson.peerDependencies || {},
+        scripts: packageJson.scripts || {},
+        name: packageJson.name || '',
+        version: packageJson.version || '',
+        description: packageJson.description || '',
+        success: true,
+      };
+    } catch (error) {
+      console.warn(`Failed to read template package.json: ${error instanceof Error ? error.message : String(error)}`);
+      return {
+        dependencies: {},
+        devDependencies: {},
+        peerDependencies: {},
+        scripts: {},
+        name: '',
+        version: '',
+        description: '',
+        success: true, // This is a graceful fallback, not a failure
+      };
+    }
+  },
+});
+
+// Step 3: Discover template units by scanning the templates directory
+const discoverUnitsStep = createStep({
+  id: 'discover-units',
+  description: 'Discover template units by analyzing the templates directory structure',
+  inputSchema: CloneTemplateResultSchema,
+  outputSchema: DiscoveryResultSchema,
+  execute: async ({ inputData, runtimeContext }) => {
+    const { templateDir } = inputData;
+
+    const tools = await AgentBuilderDefaults.DEFAULT_TOOLS(templateDir);
+
+    try {
+      const agent = new Agent({
+        model: resolveModel(runtimeContext),
+        instructions: `You are an expert at analyzing Mastra projects.
+
+Your task is to scan the provided directory and identify all available units (agents, workflows, tools, MCP servers, networks).
+
+Mastra Project Structure Analysis:
+- Each Mastra project has a structure like: ${AgentBuilderDefaults.DEFAULT_FOLDER_STRUCTURE.agent}, ${AgentBuilderDefaults.DEFAULT_FOLDER_STRUCTURE.workflow}, ${AgentBuilderDefaults.DEFAULT_FOLDER_STRUCTURE.tool}, ${AgentBuilderDefaults.DEFAULT_FOLDER_STRUCTURE['mcp-server']}, ${AgentBuilderDefaults.DEFAULT_FOLDER_STRUCTURE.network}
+- Analyze TypeScript files in each category directory to identify exported units
+
+CRITICAL: YOU MUST USE YOUR TOOLS (readFile, listDirectory) TO DISCOVER THE UNITS IN THE TEMPLATE DIRECTORY.
+
+IMPORTANT - Agent Discovery Rules:
+1. **Multiple Agent Files**: Some templates have separate files for each agent (e.g., evaluationAgent.ts, researchAgent.ts)
+2. **Single File Multiple Agents**: Some files may export multiple agents (look for multiple 'export const' or 'export default' statements)
+3. **Agent Identification**: Look for exported variables that are instances of 'new Agent()' or similar patterns
+4. **Naming Convention**: Agent names should be extracted from the export name (e.g., 'weatherAgent', 'evaluationAgent')
+
+For each Mastra project directory you analyze:
```

Comment:
I think this will fail if the structure of the repo is slightly different than what we expect